test/test-utils.js

```diff
@@ -8,6 +34,82 @@ function base64UrlEncode(str) {
   ;
 }
 
+/**
+ * Verify a JWT, ensuring that the asynchronous and synchronous calls to `verify` have the same result
+ * @param {string} jwtString The JWT as a string
+ * @param {string} secretOrPrivateKey The shared secret or private key
+ * @param {object} options Verify options
+ * @param {function(err, token):void} callback
+ */
+function verifyJWTHelper(jwtString, secretOrPrivateKey, options, callback) {
+  // freeze the time to ensure the clock remains stable across the async and sync calls
+  const fakeClock = sinon.useFakeTimers({now: Date.now()});
+  let error;
+  let syncVerified;
+  try {
+    syncVerified = jwt.verify(jwtString, secretOrPrivateKey, options);
+  }
+  catch (err) {
+    error = err;
+  }
+  jwt.verify(jwtString, secretOrPrivateKey, options, (err, asyncVerifiedToken) => {
+    try {
+      if (error) {
+        expectEqualError(err, error);
+        callback(err);
+      }
+      else {
+        expect(syncVerified, 'Async/Sync token equality').to.deep.equal(asyncVerifiedToken);
+        callback(null, syncVerified);
+      }
+    }
+    finally {
+      if (fakeClock) {
+        fakeClock.restore();
```

Comment:
since the fake clock is to have the same date in both verifications, why don't we run the restore first line in this function? so that, when we call any callback the clock is the native one. Otherwise, I think we can have race conditions when the original test plays with the fake timers too 🤔

Reply:
Good point on keeping the existing fake timer in the callback. Also related to the conversation, I did experiment with having multiple fake timers From what I can determine using nested calls to the Sinon fake timers is fine. The first fake timer is restored by the second fake timer's `restore` call.

I absolutely agree on having the timer restored before the callback is called.

Reply:
Addressed in https://github.com/auth0/node-jsonwebtoken/pull/523/commits/731a76db5b78c21c73ff53a514bdb9c8081432b4 before commit was squashed.